### PR TITLE
Cmake improvements

### PR DESCRIPTION
**Pull request by piponazo**
In this PR I am proposing the addition of some improvements in the CMake code. The motivation of my work is to create a conan package (it will be hosted [here](https://github.com/piponazo/conan-libheifh)) so that we can consume **libheif** in exiv2 (see the related PR [here](https://github.com/Exiv2/exiv2/pull/1044)).

In case you do not know about [conan](https://conan.io/), it is C++ Package Manager which makes very easy to create packages of libraries for different compilers, operative systems, compilation modes, etc. My plan is to create a conan package for libheif which could be used by any other project to consume libheif in a straightforward way. 

These are the main changes I did in the CMake code (ordered by relevance):

1. Addition of CMake installation commands. With this addition, we can run `make install` , `ninja install` or similar install commands for Visual Studio, XCode, etc. 
2. Export of Package Config file so that the project can be consumed easily by other CMake projects by doing `find_package(heif)`.
3. Usage of the variables `PROJECT_NAME` and `PROJECT_VERSION` which are defined by the CMake command `project()`. I think it might be a bit controversial one of the changes I did here ... I notice that you were using different library names depending on the platform ... which add lot of complexity when trying to consume `libheif` from other cross-platform projects. 
4. Modernisation of some CMake code. Updated required CMake version from 2.8 to 3.3 (which was released in July 23, 2015). This is the version we use in Exiv2, and most of the linux distributions have that version or higher. 
5. Generate `heif_version.h` in the build directory to not pollute the source directory.

Anyways, I tried to separate all the individual changes in separated commits to make the review process easier.

### CODE REVIEW

File: libheif/CMakeLists.txt [hunk 1]
```diff
@@ -1,84 +1,98 @@
...
-endif(CMAKE_SYSTEM_PROCESSOR STREQUAL "x86_64")
+write_basic_package_version_file(heifConfigVersion.cmake COMPATIBILITY ExactVersion)
+
+install(TARGETS ${PROJECT_NAME} EXPORT heifConfig
+    RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
+    LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+    ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
+)
+
+install(FILES ${libheif_headers} DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/heif)
```
Comment (piponazo) [inline libheif/CMakeLists.txt]: question: Should the header files be installed in a folder named `heif` or `libheif` ? I would not like to add inconsistencies between the CMake and the autotools/autoconfigure mode.

File: libheif/heif_version.h.in [hunk 2]
```diff
@@ -31,6 +31,6 @@
 #define LIBHEIF_NUMERIC_VERSION @NUMERIC_VERSION@
 
 /* Version string */
-#define LIBHEIF_VERSION "@PACKAGE_VERSION@"
+#define LIBHEIF_VERSION "@PROJECT_VERSION@"
```
Comment (piponazo) [inline libheif/heif_version.h.in]: Here I was not sure if this would conflict with the `configure.ac`. I am happy to revert this change if it is conflicting.

File: libheif/CMakeLists.txt (hunk 1 above)
Comment (farindk) [reply libheif/CMakeLists.txt]: The headers should be installed in a `libheif` directory.

**Comment by farindk** [comment]
Thanks for this work. It's much appreciated.

File: libheif/CMakeLists.txt (hunk 1 above)
Comment (piponazo) [reply libheif/CMakeLists.txt]: I changed this in the latest commits. I noticed while testing with the conan package that indeed I needed `libheif`

**Comment by piponazo** [comment]
Heads up: sorry if I keep pushing some minor commits. I am testing the changes in several platforms now with the conan package that I am creating:

https://github.com/piponazo/conan-libheifh

I will notify you when I have the package working for gcc, clang, msvc on different platforms (linux, mac, windows).

**Comment by cgilles** [comment]
Note: libde265 needs a similar CMake improvement patch. As libheif team maintain libde265, please considerate to backport this PR too into libde265. Thanks in advance...

**Comment by piponazo** [comment]
> Note: libde265 needs a similar CMake improvement patch. As libheif team maintain libde265, please considerate to backport this PR too into libde265. Thanks in advance...

Sure, I'll take a look there too :wink:

**Comment by piponazo** [comment]
I finally managed to have a conan recipe which is building for the classical combination of compilers and OSs used in conan:

https://travis-ci.org/piponazo/conan-libheifh/builds/600339942
https://ci.appveyor.com/project/piponazo/conan-libheifh

Once this PR gets merged, I will point the conan recipe to the upstream libheif repository and provide binary packages in bintray.

**Comment by D4N** [comment]
The travis error appears to be a fluke: @farindk can you restart the one failing job?

File: CMakeLists.txt [hunk 3]
```diff
@@ -61,12 +64,6 @@ endif()
 
 include_directories ("${PROJECT_SOURCE_DIR}")
 
-if(UNIX)
-  set(LIBHEIF_LIBRARY_NAME heif)
```
Comment (fancycode) [inline CMakeLists.txt]: This is necessary so the resulting file on Linux is named `libheif.so`. Without it, the name is `liblibheif.so`:
https://travis-ci.org/strukturag/libheif/jobs/599715352#L664
```
[ 43%] Linking CXX shared library liblibheif.so
```

File: libheif/heif_version.h.in (hunk 2 above)
Comment (fancycode) [reply libheif/heif_version.h.in]: Autoconf is using `PACKAGE_VERSION`, so please keep this.

File: libheif/CMakeLists.txt [hunk 4]
```diff
@@ -1,84 +1,106 @@
...
+    )
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${X265_CFLAGS}")
+
+    if (NOT "${X265_LIBRARY_DIRS}" STREQUAL "")
+        set(X265_LINKDIR "-L${X265_LIBRARY_DIRS}")
+    endif()
+
+    target_link_libraries(${PROJECT_NAME} PRIVATE ${X265_LIBRARIES} ${X265_LINKDIR})
+
+    install(FILES heif_encoder_x265.h DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/${PROJECT_NAME})
```
Comment (fancycode) [inline libheif/CMakeLists.txt]: The header `heif_encoder_x265.h` is only used internally and doesn't need to be installed.

File: libheif/CMakeLists.txt [hunk 5]
```diff
@@ -1,84 +1,106 @@
...
+
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${LIBDE265_CFLAGS}")
+
+    if (NOT "${LIBDE265_LIBRARY_DIRS}" STREQUAL "")
+        set(LIBDE265_LINKDIR "-L${LIBDE265_LIBRARY_DIRS}")
+    endif()
+
+    target_link_libraries(${PROJECT_NAME} PRIVATE ${LIBDE265_LIBRARIES} ${LIBDE265_LINKDIR})
+
+    install(FILES heif_decoder_libde265.h DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/${PROJECT_NAME})
```
Comment (fancycode) [inline libheif/CMakeLists.txt]: The header `heif_decoder_libde265.h` is only used internally and doesn't need to be installed.

File: libheif/CMakeLists.txt [hunk 6]
```diff
@@ -1,84 +1,106 @@
...
+    ${libheif_headers}
+)
+
+# Needed to find libheif/heif_version.h while compiling the library
+target_include_directories(${PROJECT_NAME} PRIVATE ${PROJECT_BINARY_DIR})
 
-# set(CMAKE_VERBOSE_MAKEFILE on)
+# Propagate include/libheif to consume the headers from other projects
+target_include_directories(${PROJECT_NAME} PUBLIC
+    $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include/${PROJECT_NAME}>
```
Comment (fancycode) [inline libheif/CMakeLists.txt]: Will this add both `include` and `include/libheif` to the exported include path? It should be sufficient to add just `include` as we want other apps to explicitly include using `#include <libheif/heif.h>` (instead of just `#include <heif.h>`).

File: libheif/CMakeLists.txt [hunk 7]
```diff
@@ -1,84 +1,106 @@
...
-endif(CMAKE_SYSTEM_PROCESSOR STREQUAL "x86_64")
+write_basic_package_version_file(${PROJECT_NAME}ConfigVersion.cmake COMPATIBILITY ExactVersion)
+
+install(TARGETS ${PROJECT_NAME} EXPORT ${PROJECT_NAME}Config
+    RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
+    LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+    ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
+)
+
+install(FILES ${libheif_headers} DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/${PROJECT_NAME})
```
Comment (fancycode) [inline libheif/CMakeLists.txt]: This will install all headers including the various private headers. Public headers are only `heif_cxx.h`, `heif.h`, `heif_plugin.h` and `heif_version.h` (generated). See https://github.com/strukturag/libheif/blob/ae12405f9d9e69791c858f79341355e068302ffe/libheif/Makefile.am#L76

**Comment by fancycode** [comment]
> The travis error appears to be a fluke: @farindk can you restart the one failing job?

Just restarted the failing jobs.

File: CMakeLists.txt (hunk 3 above)
Comment (piponazo) [reply CMakeLists.txt]: This is due to the internal CMake variable `CMAKE_SHARED_LIBRARY_PREFIX` (whose default value for UNIX environments is `lib`: 

https://cmake.org/cmake/help/latest/variable/CMAKE_SHARED_LIBRARY_PREFIX.html

Would it be fine for you to change the clear the value of that variable on UNIX so that we end up having `libheif.so` on UNIX and `libheif.lib` on Windows ? 

My objective was to remove `LIBHEIF_LIBRARY_NAME` because we already have one variable (`PROJECT_NAME`) containing the information we need.

File: libheif/CMakeLists.txt (hunk 6 above)
Comment (piponazo) [reply libheif/CMakeLists.txt]: Good catch. I copy-pasted-modified this snippet from other project where I needed to do that and I forgot to cleanup here.

File: libheif/CMakeLists.txt (hunk 7 above)
Comment (piponazo) [reply libheif/CMakeLists.txt]: Thanks, I will adapt it ASAP

File: CMakeLists.txt (hunk 3 above)
Comment (fancycode) [reply CMakeLists.txt]: I'm fine with whatever works :smile: However if you clear `CMAKE_SHARED_LIBRARY_PREFIX`, make sure the `libpixbufloader-heif.so` from folder `gdk-pixbuf` still has the `lib` prefix.

Comment (piponazo) [reply CMakeLists.txt]: Ha, I did not take into account that target. I think I will revert then this change and come back to the original implementation.

**Comment by piponazo** [comment]
>> The travis error appears to be a fluke: @farindk can you restart the one failing job?

> Just restarted the failing jobs.

I noticed that the failing  travis-ci jobs are also failing in other branches. It seems that the failures are not related to the changes I am doing here. Am I right?

File: CMakeLists.txt (hunk 3 above)
Comment (piponazo) [reply CMakeLists.txt]: I take back what I just said. If we have different names for the target for different platforms, it is going to be a mess to consume **libheif** from other multi-platform projects. 

I will think about this and come with a proposal.

**Comment by fancycode** [comment]
> > > The travis error appears to be a fluke: @farindk can you restart the one failing job?
> 
> > Just restarted the failing jobs.
> 
> I noticed that the failing travis-ci jobs are also failing in other branches. It seems that the failures are not related to the changes I am doing here. Am I right?

Yes, this looks like a Travis hiccup. I can try to restart the jobs later.

**Comment by piponazo** [comment]
> Yes, this looks like a Travis hiccup. I can try to restart the jobs later.

Note that in master, those failures are already there:
https://travis-ci.org/strukturag/libheif/builds/597687977

File: CMakeLists.txt (hunk 3 above)
Comment (piponazo) [reply CMakeLists.txt]: I just made a change related to that. I renamed the `project()` to `heif` so that it will produce:

- Unix: libheif.so
- Windows: heif.lib & heif.dll

Then I hardcoded the include directory to `libheif` so that consumers of the library does not need to change anything to find the header files. It will only change for windows consumers, which will need to update the library name from `libheif` to `heif`. 

From my point of view, it is legit to make this change because anyways, any CMake consumer of **libheif** should now use the CMake Config file to find the dependency, and that config file is in charge of indicating the library name and include directories. Note that it is also the default behaviour in CMake to have libXXX.so for UNIX and XXX.lib on Windows. Experiments trying to modify the default behavior end up in additional code for the maintained project, and additional difficulties for the consumers of the library.

Since this is a "breaking change" for windows users, I propose to create a new release of the library if this PR gets finally merged.

**Comment by piponazo** [comment]
Hi, do I need to do something else about this PR? I would like to see this contribution included into the project soon to unblock some work we are doing in Exiv2 (https://github.com/Exiv2/exiv2/pull/1044)

File: libheif/CMakeLists.txt [hunk 8]
```diff
@@ -89,3 +89,20 @@ if(X265_FOUND)
     target_link_libraries(${PROJECT_NAME} ${X265_LIBRARIES} ${X265_LINKDIR})
 endif()
 
+# Install commands
+# --------------------------
+write_basic_package_version_file(heifConfigVersion.cmake COMPATIBILITY ExactVersion)
+
+install(TARGETS ${PROJECT_NAME} EXPORT heifConfig
```
Comment (cryptomilk) [inline libheif/CMakeLists.txt]: Please use `${PROJECT_NAME}-config`

File: libheif/CMakeLists.txt [hunk 9]
```diff
@@ -89,3 +89,20 @@ if(X265_FOUND)
...
+    RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
+    LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+    ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
+)
+
+install(FILES ${libheif_headers}
+    DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/heif)
+
+install(EXPORT heifConfig DESTINATION "${CMAKE_INSTALL_LIBDIR}/cmake/heif")
```
Comment (cryptomilk) [inline libheif/CMakeLists.txt]: Please use `${PROJECT_NAME}-config` and the DESTINATION should be: `${CMAKE_INSTALL_LIBDIR}/cmake/${PROJECT_NAME}`

File: libheif/CMakeLists.txt [hunk 10]
```diff
@@ -89,3 +89,20 @@ if(X265_FOUND)
     target_link_libraries(${PROJECT_NAME} ${X265_LIBRARIES} ${X265_LINKDIR})
 endif()
 
+# Install commands
+# --------------------------
+write_basic_package_version_file(heifConfigVersion.cmake COMPATIBILITY ExactVersion)
```
Comment (cryptomilk) [inline libheif/CMakeLists.txt]: I would prefer `${PROJECT_NAME}-config-version.cmake`

File: libheif/CMakeLists.txt [hunk 11]
```diff
@@ -89,3 +89,20 @@ if(X265_FOUND)
...
+    LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+    ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
+)
+
+install(FILES ${libheif_headers}
+    DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/heif)
+
+install(EXPORT heifConfig DESTINATION "${CMAKE_INSTALL_LIBDIR}/cmake/heif")
+
+install(FILES ${CMAKE_CURRENT_BINARY_DIR}/heifConfigVersion.cmake DESTINATION "${CMAKE_INSTALL_LIBDIR}/cmake/heif")
```
Comment (cryptomilk) [inline libheif/CMakeLists.txt]: See above :-) Thanks!

File: libheif/CMakeLists.txt (hunk 8 above)
Comment (piponazo) [reply libheif/CMakeLists.txt]: Right now I am using `${PROJECT_NAME}Config` which also works. Note that the CMake documentation mentions that both format are accepted:

https://cmake.org/cmake/help/v3.16/command/find_package.html#full-signature-and-config-mode

Is there some limitation about the current name you are aware of?

File: libheif/CMakeLists.txt (hunk 9 above)
Comment (piponazo) [reply libheif/CMakeLists.txt]: In the latest version I am already using `${CMAKE_INSTALL_LIBDIR}/cmake/${PROJECT_NAME}` (I noticed the usage of hard-coded strings :smile_cat: ) .

File: libheif/CMakeLists.txt (hunk 10 above)
Comment (piponazo) [reply libheif/CMakeLists.txt]: I just would like to know if it is just a "preference" or is there some real limitation with the current naming. But I will change it anyways, since I do not have a strong opinion about this point. I am just asking to learn :wink:

Comment (cryptomilk) [reply libheif/CMakeLists.txt]: It is more a preference using the dashes than the camelcase.

Comment (piponazo) [reply libheif/CMakeLists.txt]: Alright, I just wanted to be sure I was not missing something. I already pushed the changes 😉

File: libheif/CMakeLists.txt [hunk 12]
```diff
@@ -68,12 +68,14 @@ endif()
 add_definitions(-DHAVE_VISIBILITY)
 add_definitions(-DLIBHEIF_EXPORTS)
 
-add_library(${LIBHEIF_LIBRARY_NAME} SHARED ${libheif_sources})
+add_library(${PROJECT_NAME} SHARED ${libheif_sources})
```
Comment (cryptomilk) [inline libheif/CMakeLists.txt]: I think this is wrong, as the PROJECT_NAME is libheif which would result in liblibheif. In the meantime you do:

```
add_library(heif ${libheif_sources})
target_link_libraries(heif ...)

add_library(heif::heif ALIAS heif)
```
If you want to link against heif, use the alias `heif::heif`. This way you can also modify it.

instead of add_library(... SHARED ...) you remove the shared and add:

```
option(BUILD_SHARED_LIBS "Build shared libraries" ON)
```

Then you can easily build as a static lib too. This is what the documentation for BUILD_SHARED_LIBS suggest.

Comment (piponazo) [reply libheif/CMakeLists.txt]: Did you read the comment I posted here?
https://github.com/strukturag/libheif/pull/168/files#r337861735

The original name of the project was `libheif` and I wanted to change as little as possible to avoid having a bigger impact in codebases consuming **libheif**. I also think it makes sense to use `${PROJECT_NAME}`  as the library target name to avoid hardcoded values. 

I think I did not understand you completely about your point with the line
```
add_library(heif::heif ALIAS heif)
```
In case we would want to add a namespace for the library target, should not we use the NAMESPACE option in the `install(EXPORT ...)` command?
https://cmake.org/cmake/help/v3.13/command/install.html#installing-exports

Finally, I completely agree with the `BUILD_SHARED_LIBS` suggestion. It is also something I normally use in my projects. I will add it.

**Comment by fancycode** [comment]
> The failures in TravisCI seems to be due to things unrelated to the changes I did here.

@piponazo The CI issues have been fixed on master, please rebase your branch to get the latest changes.

File: libheif/CMakeLists.txt (hunk 12 above)
Comment (piponazo) [reply libheif/CMakeLists.txt]: I just pushed to additional commit adding the support for `BUILD_SHARED_LIBS` and other one changing the name of the project to `heif` (more information in the commit message).

**Comment by piponazo** [comment]
Hi. Any news about this? It seems that CI is green and there are not conflicts. I will like to get this merged to unblock some work in Exiv2.

**Comment by farindk** [comment]
I have integrated your changes into the `piponazo-cmakeImprovements` branch.

I adapted it a bit to get rid of the NUMERIC_VERSION in the CMakeLists and also in the configure.ac.
It was not building anymore before that when using autoconf, as @PROJECT_VERSION@ was used in heif_version.h.in

**Comment by farindk** [comment]
@fancycode the branch `piponazo-cmakeImprovements` looks good to me now.
Do you want to have a look whether it still builds fine during packaging before I merge this into `master`?
The `autoconf` build is also affected by some of the changes above, so it may be worth to have a look.

**Comment by fancycode** [comment]
@farindk the branch `piponazo-cmakeImprovements` in our repository is a wild mix of the original commits from this branch, (seemingly) unrelated changes and commits that already exist on `master`.

I would rather finish the review here, merge the PR and then perform additional changes on master or a new PR.

The only thing that should be changed here imho is the use of `@PROJECT_VERSION@` in `libheif/heif_version.h.in` which I fixed in 9cb36a4b3ac48c7936d502ff3a0ba240f10c3baf.

From my point of view this is now ready to merge (once CI is green). It still builds for me with autoconf/automake and with cmake. Code looks good, too.

@piponazo Sorry for the delay and thanks for your contribution!

**Comment by farindk** [comment]
Your changes have been merged into `master`.
Thanks for the PR.
